bundles/io/org.eclipse.smarthome.io.security/src/main/java/org/eclipse/smarthome/io/security/api/TlsNetworkServer.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.security.api;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+
+import javax.net.ssl.SSLContext;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * An implementation of this interface would return an SSLContexts for a given context string,
+ * pre-configured for secure transmissions.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public interface TlsNetworkServer {
+    /**
+     * The default context constant. You will usually use null for methods of
+     * this service to acquire the default context.
+     * If you implement the {@link TLSContextChangedListener} though,
+     * you want to compare to this constant if you are interested in the default context.
+     */
+    static String DEFAULT_CONTEXT = "default";
+    static String TLS = "TLS";
+    static String DTLS = "DTLS";
+
+    /**
+     * Creates a new SSLContext with the given context or default context if the given one can't be found.
+     * This method is blocking and can take a while to resolve. If the default context is not existing yet,
+     * an implementation is required to create a self-signed certificate and public/private-key pair.
+     *
+     * @param context A context. Use {@link TlsNetworkServer#DEFAULT_CONTEXT} for a default context.
+     * @param securityProvider Either {@link TlsNetworkServer#TLS} for TCP connections or
+     *            {@link TlsNetworkServer#DTLS} for UDP connections on Java 9 and newer.
+     * @return An SSLContext, that can be used to derive a SSLEngine or a SSLServerSocketFactory.
+     * @throws IOException If loading or saving the keyStore fails or reading or writing the configuration fails, this
+     *             exception is thrown.
+     * @throws GeneralSecurityException This exception is raised if creating the self
+     *             signed certificate fails (because the requested algorithms are not supported for instance) or if the
+     *             existing certificate is expired and service is configured to not automatically create a new one.
+     */
+    SSLContext createSSLContext(String context, String securityProvider) throws GeneralSecurityException, IOException;
```

Comment:
In retrospective I would say this API is not very OSGi like. The context parameter should be removed. A service factory should be used. The requester should be able to request a special SSLContext by providing a context string via service parameter.